pkg/scheduler/schedule_one.go

```diff
@@ -213,108 +208,63 @@ func (sched *Scheduler) schedulingCycle(ctx context.Context, state *framework.Cy
 		podsToActivate.Map = make(map[string]*v1.Pod)
 	}
 
-	return scheduleResult, assumedPodInfo
+	return scheduleResult, assumedPodInfo, nil
 }
 
-// bindingCycle tries to bind an assumed Pod.
-func (sched *Scheduler) bindingCycle(ctx context.Context, state *framework.CycleState, fwk framework.Framework, scheduleResult ScheduleResult, assumedPodInfo *framework.QueuedPodInfo, podsToActivate *framework.PodsToActivate, start time.Time) {
-	assumedPod := assumedPodInfo.Pod
+func (sched *Scheduler) handleSchedulingCycleError(
+	ctx context.Context,
+	state *framework.CycleState,
+	fwk framework.Framework,
+	podInfo *framework.QueuedPodInfo,
+	start time.Time,
+	scheduleResult ScheduleResult,
+	err error) {
+
+	switch scheduleResult.reason {
+	case v1.PodReasonUnschedulable:
+		metrics.PodUnschedulable(fwk.ProfileName(), metrics.SinceInSeconds(start))
+	case v1.PodReasonSchedulerError:
+		metrics.PodScheduleError(fwk.ProfileName(), metrics.SinceInSeconds(start))
+	}
 
-	waitOnPermitStatus := fwk.WaitOnPermit(ctx, assumedPod)
-	if !waitOnPermitStatus.IsSuccess() {
-		var reason string
-		if waitOnPermitStatus.IsUnschedulable() {
-			metrics.PodUnschedulable(fwk.ProfileName(), metrics.SinceInSeconds(start))
-			reason = v1.PodReasonUnschedulable
-		} else {
-			metrics.PodScheduleError(fwk.ProfileName(), metrics.SinceInSeconds(start))
-			reason = v1.PodReasonSchedulerError
-		}
-		// trigger un-reserve plugins to clean up state associated with the reserved Pod
-		fwk.RunReservePluginsUnreserve(ctx, state, assumedPod, scheduleResult.SuggestedHost)
-		if forgetErr := sched.Cache.ForgetPod(assumedPod); forgetErr != nil {
-			klog.ErrorS(forgetErr, "scheduler cache ForgetPod failed")
-		} else {
-			// "Forget"ing an assumed Pod in binding cycle should be treated as a PodDelete event,
-			// as the assumed Pod had occupied a certain amount of resources in scheduler cache.
-			// TODO(#103853): de-duplicate the logic.
-			// Avoid moving the assumed Pod itself as it's always Unschedulable.
-			// It's intentional to "defer" this operation; otherwise MoveAllToActiveOrBackoffQueue() would
-			// update `q.moveRequest` and thus move the assumed pod to backoffQ anyways.
-			if waitOnPermitStatus.IsUnschedulable() {
-				defer sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, func(pod *v1.Pod) bool {
-					return assumedPod.UID != pod.UID
-				})
-			} else {
-				sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, nil)
-			}
+	if scheduleResult.unreserve {
```

Comment:
If we remove this from here, we can just add the metrics handling into `FailureHandler` below and remove this function.

Reply:
The problem here is if we move the callings of` UnReserve` and `ForgetPod` after `RunReservePluginsReserve`, and add the metrics to `FailureHandler`, the `metrics.SinceInSeconds(start)` is not correct, it includes the unreserve operating time. Unless we record an end time.

Reply:
The metric is `scheduling_attempt_duration_seconds` which should contain whatever is part of the cycle. So this would actually be a fix

Reply:
Fair, created a separate fixed PR here https://github.com/kubernetes/kubernetes/pull/113113.

Reply:
Moved the metrics to `FailureHandler`